examples/planetoid_train.py

```diff
@@ -0,0 +1,612 @@
+            data.num_nodes,
+        )
+    elif gnn_choice == 'pmlp':
+        model = PMLP(
+            in_channels=dataset.num_features,
+            hidden_channels=args.hidden_channels,
+            out_channels=dataset.num_classes,
+            num_layers=2,
+            dropout=0.5,
+            norm=False,
+        )
+    elif gnn_choice == 'splinegnn':
+        model = SplineGNN(
+            dataset.num_features,
+            args.hidden_channels,
+            dataset.num_classes,
+        )
+    elif gnn_choice == 'gcn2':
+        model = GCN2(
+            dataset.num_features,
+            args.hidden_channels,
+            dataset.num_classes,
+            num_layers=64,
+            alpha=0.1,
+            theta=0.5,
+            shared_weights=True,
+            dropout=0.6,
+        )
+    elif gnn_choice == 'super_gat':
+        model = SuperGAT(
+            dataset.num_features,
+            dataset.num_classes,
+        )
+    elif gnn_choice == 'dna':
+        model = DNA(
+            dataset.num_features,
+            args.hidden_channels,
+            dataset.num_classes,
+            num_layers=5,
+            heads=8,
+            groups=16,
+        )
+    elif gnn_choice == 'gat':
+        model = GAT(
+            dataset.num_features,
+            args.hidden_channels,
+            dataset.num_classes,
+            heads=8,
+        )
+    elif gnn_choice == 'gcn':
+        model = GCN(
+            dataset.num_features,
+            args.hidden_channels,
+            dataset.num_classes,
+        )
+    elif gnn_choice == 'node2vec':
+        model = Node2Vec(
+            data.edge_index,
+            embedding_dim=args.hidden_channels,
+            walk_length=20,
+            context_size=10,
+            walks_per_node=10,
+            num_negative_samples=1,
+            p=1.0,
+            q=1.0,
+            sparse=True,
+        )
+    else:
+        raise ValueError(f'Unsupported model type: {gnn_choice}')
+    return model
+
+
+def get_optimizer(
+    gnn_choice: str,
+    model: torch.nn.Module,
+) -> torch.optim.Optimizer:
+    if gnn_choice == 'node2vec':
+        optimizer = torch.optim.SparseAdam(model.parameters(), lr=args.lr)
+    else:
+        optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,
+                                     weight_decay=args.wd)
+    return optimizer
+
+
+model = get_model(args.gnn_choice).to(device)
+optimizer = get_optimizer(args.gnn_choice, model)
+if args.gnn_choice == 'node2vec':
+    num_workers = 4 if sys.platform == 'linux' else 0
+    loader = model.loader(batch_size=128, shuffle=True,
+                          num_workers=num_workers)
+
+
+def train_node2vec():
+    model.train()
+    total_loss = 0
+    for pos_rw, neg_rw in loader:
+        optimizer.zero_grad()
+        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
+        loss.backward()
+        optimizer.step()
+        total_loss += loss.item()
+    return total_loss / len(loader)
```

Comment:
Let's accumulate the loss as a tensor and not as a float so that we don't need to cuda sync every iteration:
```suggestion
    total_loss = torch.zeros(1, device=device).squeeze()
    for pos_rw, neg_rw in loader:
        optimizer.zero_grad()
        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
        loss.backward()
        optimizer.step()
        total_loss += loss
    return total_loss.item() / len(loader)
```